lib/data/model.ts

```diff
@@ -169,6 +169,10 @@ export default class Model extends DenaliObject {
     return new Proxy(this, {
 
       get(model: Model, property: string): any {
+        if (property === 'constructor') {
```

Comment:
So maybe it's not, I think what I was actually running into was the `model.adapter.getAttribute` that was messing things up. The default memory adapter simply returns `model.record[property]`, and so if your underlying record is an instance of a class, then you get the underlying record constructor instead of the model class constructor.